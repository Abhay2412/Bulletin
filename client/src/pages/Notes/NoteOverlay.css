.overlay {
    position: absolute;
    display: flex;
    height: 95vh;
    width: 98vw;
    top: 0;
    align-items: center;
    justify-content: center;
    background-color: rgba(83,64,41,0.2);
    margin: 0 !important;
    z-index: 1; 
}

.overlayBox {
    width: 80%;
    height: 60%;
    background-color: white;
    text-align: left;
    padding-left: 5%;
    padding-top: 2%;
    color: #B8A590;
}

.overlayFields {
    color: #675443;
    padding-left: 5%;
    padding-top: 2%;
}

.overlayFieldRow {
    display: flex;
    justify-content: space-between;
    padding-right: 50%;
}

.saveDiv {
    display: flex;
    justify-content: right;
    padding-top: 5%;
    padding-right: 10%;
}

.saveButton {
    background-color: #AA896B;
    color: white;
    width: 30%;
}