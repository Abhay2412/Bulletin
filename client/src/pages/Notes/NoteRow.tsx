import {Box, Button, Typography, TextField} from "@mui/material"
import {useState} from 'react'
import React from 'react'
import { StringLiteralLike } from "typescript";

interface Props {
    id: string;
    name: string;
    date: string;
    content: string;
    onDelete: (arg0: string)=>void;
    onEdit: (arg0: string, arg1: string, arg2: string, arg3: string)=>void;
}
interface NewState {
  name: string;
  content: string;
  date: string;
}

export default function NoteRow({name, date, content, id, onDelete, onEdit}: Props) {
    const [popupEditState, setPopupEditState] = useState(false);
    const [popupOpenState, setPopupOpenState] = useState(false);
    const [newValues, setNewValues] = React.useState<NewState>({
      name: name,
      content: content,
      date: date,
    });
  const handleChange =
  (prop: keyof NewState) => (event: React.ChangeEvent<HTMLTextAreaElement>) => {
    setNewValues({ ...newValues, [prop]: event.target.value });
  };
  const handleSave = () =>{
    setPopupEditState(false);
    onEdit(id, newValues.name, newValues.content, newValues.date);
  }
  return (
    <>
    {popupEditState ? (
         <div className="overlay" style={{paddingTop: "15%"}}>
         <div className="overlayBox">
           <div className="noteHeader">
             <Typography variant="h6">
               Note Name
             </Typography>
             <TextField  variant="standard" defaultValue={name} onChange={handleChange("name")}></TextField>
             <Button onClick={()=>setPopupEditState(false)}>
               <Typography variant="h5">
                   X
               </Typography>
             </Button>
           </div>
           <div className="noteContent">
               <textarea defaultValue={content} onChange={handleChange("content")}></textarea>
           </div>
           <div className="saveDiv">
                    <Button className="saveButton" onClick={handleSave}>
                        Save Note
                    </Button>
           </div>
         </div>
     </div>
    ) : null}

    {popupOpenState ? (
            <div className="overlay" style={{paddingTop: "20%"}}>
                <div className="overlayBox">
                  <div className="noteHeader">
                    <Typography variant="h4">
                      {name}
                    </Typography>
                    <Button onClick={()=>setPopupOpenState(false)}>
                    <Typography variant="h5">
                        X
                    </Typography>
                    </Button>
                  </div>
                  <div className="noteContent">
                    {content}
                  </div>
                </div>
            </div>
        ) : null}

      <Box className="notesTableRow tableCellsFormatting noteNameCell" width="70%">
        <Box>
          {name}
        </Box>
        <Box>
          <Button variant="outlined" onClick={()=>setPopupOpenState(true)}>
            Open
          </Button>
        </Box>
      </Box>
      <Box className="notesTableRow tableCellsFormatting" width="20%">
        {date}
      </Box>
      
      <Box className="notesTableRow" width="5%">
        <Button onClick={()=>onDelete(id)}>X</Button>
      </Box>
      <Box className="notesTableRow" width="5%">
        <Button sx={{ textDecoration: "underline" }} onClick={()=>setPopupEditState(true)}>edit</Button>
      </Box> 
    </>
  )
}

