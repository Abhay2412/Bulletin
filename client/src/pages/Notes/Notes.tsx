import React from "react";
import { Container, Typography, TextField, Button, Box } from "@mui/material";
import { useEffect, useState } from "react";
import NoteRow from "./NoteRow";
import "./Notes.css";
import { useParams } from "react-router-dom";
import axiosInstance from "../../axios";
import { Snackbar } from "@material-ui/core";
import { Alert, AlertColor } from "@mui/material";
import { Note } from "../../../../types";
import { format } from "date-fns";

interface CreateNoteErrors {
  name: string;
  content: string;
}

const Notes: React.FC = () => {
  const params = useParams();
  const [popupState, setPopupState] = useState(false);
  const [boardName, setBoardName] = useState("");
  const [notes, setNotes] = useState<Array<Note>>([]);
  const [noteAdded, setNoteAdded] = useState(false);

  const [values, setValues] = React.useState<Note>({
    id: null,
    author: "Author",
    timestamp: parseInt(format(new Date(), "T")),
    content: "content",
  });
  const [errors, setErrors] = useState<CreateNoteErrors>({
    name: "",
    content: "",
  });
  const [message, setMessage] = useState("");
  const [messageOpen, setMessageOpen] = useState(false);
  const [messageSeverity, setMessageSeverity] = useState<AlertColor>("success");
  const handleChange =
    (prop: keyof Note) => (event: React.ChangeEvent<HTMLTextAreaElement>) => {
      setValues((prev) => ({
        ...prev,
        [prop]: event.target.value,
        timestamp: parseInt(format(new Date(), "T")),
      }));
    };

  useEffect(() => {
    let success = true;
    axiosInstance
      .get("/readNotes", { params: { boardID: params.board_id } })
      .then((res) => {
        console.log(res);
        if (success) {
          setNotes(res.data.content);
        }
      })
      .catch((err) => {
        console.log("error getting user boards: ", err);
        success = false;
      });
  }, [noteAdded]);

  useEffect(() => {
    axiosInstance
      .get("/readBoard", { params: { boardID: params.board_id } })
      .then((res) => {
        setBoardName(res.data.content.name);
        console.log("Information recieved Successfully");
      })
      .catch((err) => {
        console.log("error getting user boards: ", err);
      });
  }, [params.board_id]);

  const validateData = () => {
    errors.name = "";
    errors.content = "";

    let errorsExits = false;
    if (!values.author) {
      errors.name = "Please enter a valid note Name.";
      errorsExits = true;
    }

    if (!values.content) {
      errors.content = "Please enter a valid note content.";
      errorsExits = true;
    }

    setErrors({ ...errors });
    return !errorsExits;
  };

  const handleEditNote = (
    id: string,
    newName: string,
    newContent: string,
    newDate: number
  ) => {
    console.log(id);
    console.log(newName);
    console.log(newContent);
    let success = true;

    const note: Note = {
      author: newName,
      content: newContent,
      id: id,
      timestamp: newDate,
    };

    axiosInstance
      .put("/updateNote", note, {
        params: { noteID: id, boardID: params.board_id },
      })
      .then((res) => {
        console.log(res);
        if (success) {
          setMessage("Note edited.");
          setMessageSeverity("success");
          setNoteAdded(!noteAdded);
        } else {
          setMessage("Note cannot be edited.");
          setMessageSeverity("error");
        }
      })
      .catch((err) => {
        console.log("error deleting note: ", err);
        success = false;
      });
  };

  const handleDeleteNote = (id: string) => {
    console.log(id);
    let success = true;
    axiosInstance
      .delete("/deleteNote", {
        params: { noteID: id, boardID: params.board_id },
      })
      .then((res) => {
        console.log(res);
        if (success) {
          setMessage("Note deleted.");
          setMessageSeverity("success");
          setNoteAdded(!noteAdded);
        } else {
          setMessage("Note cannot be deleted.");
          setMessageSeverity("error");
        }
      })
      .catch((err) => {
        console.log("error deleting note: ", err);
        success = false;
      });
  };

  const handleSaveNote = () => {
    if (!validateData()) {
      return;
    }
    let newDate = new Date();

    // values.timestamp = parseInt(format(newDate, "T"));
    console.log(params.board_id);
    console.log(values);
    let success = true;

    const newNote: Note = {
      id: null,
      content: values.content,
      timestamp: parseInt(format(newDate, "T")),
      author: values.author,
    };

    axiosInstance
      .post("/createNote", newNote, { params: { boardID: params.board_id } })
      .then((res) => {
        console.log(res);
        if (success) {
          setMessage("Note Added.");
          setMessageSeverity("success");
          setNoteAdded(!noteAdded);
        } else {
          setMessage("Note cannot be added.");
          setMessageSeverity("error");
        }
        setValues({
          id: null,
          author: "Author",
          timestamp: 1000,
          content: "content",
        });
        setPopupState(false);
      })
      .catch((err) => {
        console.log("error adding note: ", err);
        success = false;
      });
  };

  return (
    <Container sx={{ height: "100vh" }}>
      {popupState ? (
        <div className="overlay">
          <div className="overlayBox">
            <div className="noteHeader">
              <Typography variant="h6">New Note</Typography>
              <TextField
                variant="standard"
                defaultValue={values.author}
                onChange={handleChange("author")}
                helperText={`${errors.name}`}
                error={errors.name != ""}
              />
              <Button onClick={() => setPopupState(false)}>
                <Typography variant="h5">X</Typography>
              </Button>
            </div>

            <TextField
              className="InputText"
              variant="outlined"
              value={values.content}
              onChange={handleChange("content")}
              focused
              align-items="left"
              id="description-text"
              rows="8"
              helperText={`${errors.content}`}
              multiline
              style={{ fontWeight: "bold", marginTop: "10px" }}
              error={errors.content !== ""}
            />

            <div className="saveDiv">
              <Button className="saveButton" onClick={handleSaveNote}>
                Save Note
              </Button>
            </div>
          </div>
        </div>
      ) : null}

      <a href={"/board/" + params.board_id}>
        <Typography
          variant="subtitle1"
          color="primary.light"
          align="left"
          sx={{ pt: "1vh" }}
        >
          Back to Board {boardName} - Main
        </Typography>
      </a>

      <Typography className="notesTitle" variant="h3">
        Notes
      </Typography>

      <Box>
        <Box className="notesTable">
          <Box className="notesTableRow">
            <Box width={"70%"}>
              <Typography fontWeight={"bold"}>Note Title</Typography>
            </Box>
            <Box width={"20%"}>
              <Typography fontWeight={"bold"}>Date Created</Typography>
            </Box>

            <Box width={"5%"}></Box>
            <Box width={"5%"}></Box>
          </Box>
          {notes.map((notes, idx) => {
            return (
              <Box key={idx} className="notesTableRow">
                <NoteRow
                  id={notes.id}
                  author={notes.author}
                  timestamp={notes.timestamp}
                  content={notes.content}
                  onDelete={handleDeleteNote}
                  onEdit={handleEditNote}
                ></NoteRow>
              </Box>
            );
          })}
          <Box className="notesTableRow tableCellsFormatting addNewBox">
            <Button
              variant="text"
              fullWidth={true}
              onClick={() => setPopupState(true)}
            >
              <Typography>+ Add New Note</Typography>
            </Button>
          </Box>
        </Box>
      </Box>
      <Snackbar
        open={messageOpen}
        autoHideDuration={6000}
        onClose={() => setMessageOpen(false)}
      >
        <Alert severity={messageSeverity} sx={{ width: "100%" }}>
          {message}
        </Alert>
      </Snackbar>
    </Container>
  );
};

export default Notes;
