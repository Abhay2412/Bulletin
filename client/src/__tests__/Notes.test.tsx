import React, { useState } from "react";
import Notes from "../pages/Notes/Notes";
import { configure, mount, ReactWrapper } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import { store } from "../store";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";

configure({ adapter: new Adapter() });

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    board_id: "g7HGuvfnGe3nhQgAF8qY",
  }),
}));

describe("Testing <Notes /> Component", () => {
  it("should open a modal", () => {
    const wrapper: ReactWrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <Notes />
        </BrowserRouter>
      </Provider>
    );

    wrapper.find("button").find("#add-note-button").simulate("click");
    expect(wrapper.find(".overlayBox").exists()).toBeTruthy();
  });

  describe("Need open modal setup", () => {
    const wrapper: ReactWrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <Notes />
        </BrowserRouter>
      </Provider>
    );

    beforeEach(() => {
      wrapper.find("button").find("#add-note-button").simulate("click");
    });

    it("should change note title value", () => {
      wrapper
        .find("input")
        .find("#add-note-name-field")
        .simulate("change", {
          target: { value: "Test Note" },
        });
      expect(
        wrapper.find("input").find("#add-note-name-field").get(0).props.value
      ).toEqual("Test Note");
    });

    it("should display note name error", () => {
      wrapper
        .find("input")
        .find("#add-note-name-field")
        .simulate("change", {
          target: { value: "" },
        });

      wrapper.find("button").find("#save-note-button").simulate("click");
      expect(
        wrapper.find("p").find("#add-note-name-field-helper-text").text()
      ).toEqual("Please enter a valid note Name.");
    });

    it("should change note description value", () => {
      wrapper
        .find("textarea")
        .find("#add-note-description-field")
        .simulate("change", {
          target: { value: "Test Description" },
        });
      expect(
        wrapper.find("textarea").find("#add-note-description-field").get(0)
          .props.value
      ).toEqual("Test Description");
    });

    it("should display note description error", () => {
      wrapper
        .find("textarea")
        .find("#add-note-description-field")
        .simulate("change", {
          target: { value: "" },
        });

      wrapper.find("button").find("#save-note-button").simulate("click");
      expect(
        wrapper.find("p").find("#add-note-description-field-helper-text").text()
      ).toEqual("Please enter a valid note content.");
    });

    it("should close modal", () => {
      wrapper.find("button").find("#close-note-modal-button").simulate("click");
      expect(wrapper.find(".overlayBox").exists()).toBeFalsy();
    });
  });

  it("should open a snackbar", () => {
    const wrapper: ReactWrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <Notes />
        </BrowserRouter>
      </Provider>
    );

    const setValue = (arg: boolean) => {
      return arg;
    };

    jest.mock("react", () => ({
      ...jest.requireActual("react"),
      useState: jest.fn(() => [true, setValue]),
    }));

    expect(wrapper.find(".MuiSnackbar-root").exists()).toBeTruthy();
  });
});
