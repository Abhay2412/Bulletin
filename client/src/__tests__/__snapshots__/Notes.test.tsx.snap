// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Notes /> Component should render notes page correctly 1`] = `
Array [
  <Context.Provider
    value={
      Object {
        "store": Object {
          "@@observable": [Function],
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "subscription": Object {
          "addNestedSub": [Function],
          "getListeners": [Function],
          "handleChangeWrapper": [Function],
          "isSubscribed": [Function],
          "notifyNestedSubs": [Function],
          "onStateChange": [Function],
          "trySubscribe": [Function],
          "tryUnsubscribe": [Function],
        },
      }
    }
  >
    <BrowserRouter>
      <Notes />
    </BrowserRouter>
  </Context.Provider>,
]
`;
